upgrades/4.1_5.0/tams-4.1-5.0-upgrade-template.yaml

```diff
@@ -0,0 +1,212 @@
+
+  UtilsLayerArn:
+    default: The UtilsLayerArn Output from the TAMS API CloudFormation Stack
+
+  WebhooksTableName:
+    default: The name of the DynamoDB table for Webhooks
+
+  OldStorageTableName:
+    default: The name of the OLD DynamoDB table for FlowStorage. The Old Table has the flow_id as the sort key
+
+  NewStorageTableName:
+    default: The name of the NEW DynamoDB table for FlowStorage. The New Table does not have a sort key
+
+Parameters:
+  LambdaSecurityGroup:
+    Type: String
+
+  LambdaSubnetIds:
+    Type: CommaDelimitedList
+
+  NeptuneClusterResourceId:
+    Type: String
+
+  NeptuneEndpoint:
+    Type: String
+
+  UtilsLayerArn:
+    Type: String
+
+  WebhooksTableName:
+    Type: String
+
+  OldStorageTableName:
+    Type: String
+
+  NewStorageTableName:
+    Type: String
+
+Transform: AWS::Serverless-2016-10-31
+
+Resources:
+  LambdaFunction:
+    Type: AWS::Serverless::Function
+    Metadata:
+      cfn_nag:
+        rules_to_suppress:
+          - id: W89
+            reason: Vpc defined in Globals section
+          - id: W92
+            reason: ReservedConcurrentExecutions not required
+    Properties:
+      Timeout: 900
+      MemorySize: 10240
+      Handler: index.lambda_handler
+      Runtime: python3.13
+      Architectures:
+        - arm64
+      Environment:
+        Variables:
+          POWERTOOLS_LOG_LEVEL: INFO
+          WEBHOOKS_TABLE: !Ref WebhooksTableName
+          OLD_STORAGE_TABLE: !Ref OldStorageTableName
+          NEW_STORAGE_TABLE: !Ref NewStorageTableName
+          NEPTUNE_ENDPOINT: !Ref NeptuneEndpoint
+      Layers:
+        - !Sub arn:${AWS::Partition}:lambda:${AWS::Region}:017000801446:layer:AWSLambdaPowertoolsPythonV3-python313-arm64:23
+        - !Ref UtilsLayerArn
+      VpcConfig:
+        SubnetIds: !Ref LambdaSubnetIds
+        SecurityGroupIds:
+          - !Ref LambdaSecurityGroup
+      Policies:
+        - Version: "2012-10-17"
+          Statement:
+            - Effect: Allow
+              Action:
+                - dynamodb:Scan
+              Resource:
+                - !Sub arn:${AWS::Partition}:dynamodb:${AWS::Region}:${AWS::AccountId}:table/${WebhooksTableName}
+                - !Sub arn:${AWS::Partition}:dynamodb:${AWS::Region}:${AWS::AccountId}:table/${OldStorageTableName}
+            - Effect: Allow
+              Action:
+                - dynamodb:BatchWriteItem
+              Resource:
+                - !Sub arn:${AWS::Partition}:dynamodb:${AWS::Region}:${AWS::AccountId}:table/${NewStorageTableName}
+            - Effect: Allow
+              Action:
+                - neptune-db:ReadDataViaQuery
+                - neptune-db:WriteDataViaQuery
+                - neptune-db:DeleteDataViaQuery
+              Resource: !Sub arn:${AWS::Partition}:neptune-db:${AWS::Region}:${AWS::AccountId}:${NeptuneClusterResourceId}/*
+              Condition:
+                StringEquals:
+                  neptune-db:QueryLanguage: OpenCypher
+      InlineCode: |
```

Comment:
For each of these methods, I don't see items being removed from the old table or persisting progress. If this is re-run subsequently is it going to re-process everything? If there's too much data to complete in the timeout might it never complete? (If that's a valid concern)

Reply:
For the webhooks, we need all the records so this cannot be run multiple times. I will add a comment on this. However, the number of webhooks are unlikely to ever exceeed the number that could be processed by Lambda in 15mins.

For the Storage records the function return the `LastEvaluatedKey` so that the next invocation can pick up from where the previous one reached.

For the Tags it is updating the record in place and the initial query will exclude those that were updated on a previous run so this does not need to track progress and will not re-run previous updates.